### ISSUE TITLE
Clonning a SceneNode 

### ISSUE THREAD

**Issue by lluisvrx** [helix-toolkit/helix-toolkit#2037]
Hi,

Is there a way of clonning a SceneNode? I would like to add the same SceneNode to a SceneNodeGroupModel3D but as I try to add the same one, with the exact same GUID, the .AddNode returns false and ignores the command! I guess if I could clone the SceneNode with a different identifier I would be able to add it to the group.

I'm using HelixToolkit.WPF.SharpDX in case it matters.

Thanks!

**Comment by holance** [comment]
There's no clone function right now. You can create your own clone function or wait for me to create one when I have spare time.

**Comment by lluisvrx** [comment]
Sure, I would be glad to help you but how could I start?

My first thought was to declare a `new SceneNode()` and just copy all the attributes, but I see that I cannot declared that way, any idea? ^^

**Comment by lluisvrx** [comment]
It took me a while but I got what I needed.

```
public GroupNode CloneSceneNode(SceneNode scene)
        {
            var group = new GroupNode();
            foreach (var item in scene.Items)
            {
                foreach (var renderable in item.Traverse())
                {
                    if (renderable is MeshNode node)
                    {
                        var mesh = new MeshNode();
                        mesh.Geometry = node.Geometry;
                        mesh.Material = node.Material;

                        group.AddChildNode(mesh);
                    }
                }
            }

            return group;
        }
```

**Comment by lluisvrx** [comment]
You are right, I was interested in clonning meshes/shapes and colors/materials, but from here I guess it's just add the missing parts of the structure! :D

**Comment by weihuajiang** [comment]
I did it in my project, following is the code, which can work in my project

        static SceneNode Clone(SceneNode node, int index = 0)
        {
            var type = node.GetType();
            var n = Activator.CreateInstance(type);
            if (node is GroupNodeBase g)
            {
                foreach (var i in g.Items)
                {
                    (n as GroupNodeBase).AddChildNode(Clone(i, index + 1));
                }
            }
            foreach (var p in type.GetProperties())
            {
                if (!p.CanWrite || !p.CanRead || p.Name == "Items" || p.Name == "Parent") continue;
                var value = p.GetValue(node);
                if (value == p.GetValue(n)) continue;
                if (value is SceneNode)
                {
                    //p.SetValue(n, Clone(value as SceneNode, index+1));
                }
                else
                    p.SetValue(n, value);
            }
            return n as SceneNode;
        }
